### DISCUSSION TITLE
RPi 4 images

### DISCUSSION THREAD

**Discussion by ghost** [Peter-van-Tol/LiteX-CNC#43] [Ideas]
I've been working on a image based on Debian Bookworm for the RPi4, needs a little polishing, but does boot and runs the sims, I've been using myself whilst playing around.
 https://drive.google.com/file/d/1Ve6LismgxL0OfYufkeU1Z0hzdWhbf_WJ/view?usp=drive_link
I was wondering if if would be an idea to put one together for Litexcnc. I'd include Linuxcnc-uspace, linuxcnc-uspace-dev (drivers prebuilt), Litexcnc, Litex & Oss-cad-suite, I think the image would come out at about a 2.5GB down load and would require 16GB (min) SD Card. Whilst not super super quick it can build firmware.

**Comment by OJthe123** [comment]
RPi4 is a 64bit arch, right?

Does LinuxCNC with LiteX-CNC work together?
It seems my LiteX-CNC connection problems have to do with 64bit system

**Comment by ghost** [comment]
Yeah 64 bit. I'm actually using the the RPI kernel sources, applying the required real time patches,building the kernel and then using the Bookworm binaries for the filesystem.

I not too sure what you mean by "Does LinuxCNC with LiteX-CNC work together?"

**Comment by OJthe123** [comment]
On my M73 Tiny PC I have read errors with LiteX-CNC and the 5A-75B board.
On Thursday we did a little bug hunt and it seems the error comes from 64bit system and buffer / memory access.
So do you use the RPi4 already successfully with LiteX-CNC and a colorlight board?

**Comment by ghost** [comment]
I've been testing on a M72e with 64 bit Bookworm, so far no issues.
Just building some test firmware on the RPi and going to give it quick run.

Like I said it was an idea, I wouldn't put anything in the wild till I tested it. The above link was just a basic image that I've been using to play around with some Spartan 6 dev boards with Mesa firmware.

**Comment by ghost** [comment]
Ok, did a test, you sure that is a working hal file ?, had to remove the mab2hal & spindle sections, as able to jog around, had to bump the servo thread up to 1250000 as I was getting latency messages. Should try with another kernel and sd card. Didn't have any read error messages, what speed are you have on the interface for the card 1gbs or 100mbs ? I'm have mine at 100mbs.

Note to self: don't try to build firmware on the RPi with 3 cores isolated. Firmware was 6 Stepegns. 6 encoders, a number of PWM and about 18 or so gpios.

I was testing on a 5A-75E v7.1

**Comment by OJthe123** [comment]
Thanks for testing.
Yes, this hal is working.
Of course you get errors when using mb2hal, because you don't have the mb2hal.ini 
Therefore the mb2hal pins were not created, which are connected to the spindle signals.

But we can try to test your hal / config / firmware on my maschine / PC if you want.
Only I have the 5A-75B v8.0

I flashed an SDcard for my RPi4. Will going to test it soon.
Are the cores isolated by default in your image?

**Reply by ghost** [reply]
I knew enough to remove those sections & the spindle.

But Honestly next time it really helps to have the HAL & INI files.....so we have a better idea of whats going on. If everything is all in one place, rather tan trying to piece things together from other posts.

I got no issues with my configs, you said you sere having issues with 64 bit. I wasn't seeing anything of the sort.

Now if you have **all** your files available for testing that you can post here, I can help. Even your json 
config would be helpful as well. Just don't tell me to go here, go there I used this and I used that.

Theres no use using a previous ini file for 3 axis when your hal is for 2.

**Comment by OJthe123** [comment]
I agree with you when the problem is in the Hal and / or Ini files....
But in my opinion my problem is not related to a LinuxCNC configuration.
If I reduce the hal to only loading the litexcnc threads I have the reading errors.
This is why I asked for tests with 64bit...

A friend, where I installed my working LiteX-CNC version is now getting read errors also. After 3 weeks of working without problems.
Just want to gather as much information as possible...

**Reply by ghost** [reply]
Yeah I get that, but it also makes those that are testing have an easier time to setup things up.

**Comment by Peter-van-Tol** [comment]
The read errors are spreading? That's not good news.

The images which you've build for the RPi are good news though! I really love the effort. The image I use is 32-bit Buster, installed with the instruction I've found [here](https://forum.linuxcnc.org/9-installing-linuxcnc/47841-installing-linuxcnc-2-8-4-on-raspbian-10-buster-tested-on-raspberry-pi-3b-pi?start=0).

With `isolcpus=2,3` this setup can still build the firmware in reasonable time. What time are you looking at for building it on Bookworm using a single core?

**Reply by ghost** [reply]
Think it was about 2.5 - 3 hours, kind of set it up and went to sleep.

Just for confirmation which branch should we be using  main or 11-external ?

**Reply by Peter-van-Tol** [reply]
`main` or use the version with `pip install litexcnc[cli]`

**Reply by Peter-van-Tol** [reply]
In my case (with 2 CPUs available) it is running in way under 1 hour....

**Reply by OJthe123** [reply]
With 4 CPUs the 5a-75b test config with encoders, stepgen, etc took about 17minutes.
maybe one could enable the CPUs just for building the firmware and then switch back to isolated CPUs for latency

**Reply by ghost** [reply]
I usually build the firmware on my desktop. Doing it on the Pi was just for a bit of fun. 
But yeah enabling all four 4 cores would definitely help.

**Comment by OJthe123** [comment]
My friend told me he could not start LCNC anymore and sends me a pic with "read error...expected 48'
It can be he changed some other things I don't know yet.
Tomorrow i meet him to check out

**Reply by Peter-van-Tol** [reply]
That's the error. Basically something writes at a wrong location, overwriting part of the `read_addresses` buffer. Due to this error the FPGA doesn't recognize the packet as a command, so it does not send back any data.

**Reply by Peter-van-Tol** [reply]
There is at this moment an fix available in #44 . After testing this fix will be merged in `main`.

**Comment by OJthe123** [comment]
started my Pi4...

did you set username and password?

**Comment by ghost** [comment]
Username: cnc
Password: linuxcnc 

cmdline.txt & config.txt are under /boot/broadcom

**Comment by OJthe123** [comment]
> I've been working on a image based on Debian Bookworm for the RPi4, needs a little polishing, but does boot and runs the sims, I've been using myself whilst playing around. https://drive.google.com/file/d/1Ve6LismgxL0OfYufkeU1Z0hzdWhbf_WJ/view?usp=drive_link I was wondering if if would be an idea to put one together for Litexcnc. I'd include Linuxcnc-uspace, linuxcnc-uspace-dev (drivers prebuilt), Litexcnc, Litex & Oss-cad-suite, I think the image would come out at about a 2.5GB down load and would require 16GB (min) SD Card. Whilst not super super quick it can build firmware.

I installed pip3, litexcnc, litex and oss-cad-suite right after flashing your image.
this is the size it uses now

```
cnc@raspberrypi:~$ df -h
Dateisystem    Größe Benutzt Verf. Verw% Eingehängt auf
udev            1,6G       0  1,6G    0% /dev
tmpfs           380M    5,8M  374M    2% /run
/dev/mmcblk0p2   15G    9,4G  4,3G   70% /
tmpfs           1,9G       0  1,9G    0% /dev/shm
tmpfs           5,0M     16K  5,0M    1% /run/lock
tmpfs           1,9G     12K  1,9G    1% /tmp
/dev/mmcblk0p1  252M     53M  199M   21% /boot/broadcom
tmpfs           380M     60K  380M    1% /run/user/1000

```

**Reply by ghost** [reply]
Cheers for that.
I'll make a new image and upload it today. Hopefully it should be a case of download, write the image and boot. Drivers will be preinstalled, I'll include a readme on the desktop, that from previous experience no one will read, hahaha.
I'll keep the Linuxcnc version as 2.9.

**Comment by ghost** [comment]
New image.
User: cnc
Pasword tuxcnc

username & password included in text file in google drive folder.

Linuxcnc 2.9 (from debian repo)
Litexcnc drivers built & installed
Litex installed
Oss-cad-suite installed
Sample firmware configs included /home/cnc/LitexCNC-configs
Libre-Office removed to save a little space
Use a 16GB SD Card
udevd rules included for usb-blaster for programming firmware wtih openFGPALoader.
venv is /home/cnc/.litexcnc
All paths setup
SPI enabled
Audio & BT disabled
Wifi enabled
2 cores isolated
Rainbow splash enabled during boot
Flash your card, boot up and go for it.

On first boot filesystem will expand to fit SD Card

Be sure to read text file on desktop.
https://drive.google.com/file/d/1PvfQPvsrcP9Ss2Zrryh35VmkFn31pA4P/view?usp=sharing

Have fun

**Reply by Peter-van-Tol** [reply]
Great job!

I would recommend to include OpenOCD with the image, as this can program the FPGA using the GPIO of the RPi. Then you don't have to get an USB-blaster, a few jumper wires will suffice.

To get the BCM chipset working with the RPi you have to compile it yourself. The standard `apt-get` won't work with GPIO. A good description on how to install can be found
[here](https://learn.adafruit.com/programming-microcontrollers-using-openocd-on-raspberry-pi/compiling-openocd). 

For my HUB75HAT I'm using a fixed set of pins, when the testing is finished, we might add a default config for this HAT.

**Comment by kzali** [comment]
Great job mate 
Just put it on a download

**Comment by OJthe123** [comment]
Thanks!
Just tested it.
flash, boot, read text file 😄 

I can start linuxcnc and it connects to the colorlight board 👍 
but in my case I had to set a ip route for the eth device. otherwise it tried to connect over wifi...

**Reply by ghost** [reply]
Running the build process via my qemu dev setup, look ok ?

Ran:
`./configure --enable-sysfsgpio --enable-bcm2835gpio`

got this

```

OpenOCD configuration summary
--------------------------------------------------
MPSSE mode of FTDI based devices        yes (auto)
ST-Link Programmer                      yes (auto)
TI ICDI JTAG Programmer                 yes (auto)
Keil ULINK JTAG Programmer              yes (auto)
ANGIE Adapter                           yes (auto)
Altera USB-Blaster II Compatible        yes (auto)
Bitbang mode of FT232R based devices    yes (auto)
Versaloon-Link JTAG Programmer          yes (auto)
TI XDS110 Debug Probe                   yes (auto)
CMSIS-DAP v2 Compliant Debugger         yes (auto)
OSBDM (JTAG only) Programmer            yes (auto)
eStick/opendous JTAG Programmer         yes (auto)
Olimex ARM-JTAG-EW Programmer           yes (auto)
Raisonance RLink JTAG Programmer        yes (auto)
USBProg JTAG Programmer                 yes (auto)
Espressif JTAG Programmer               yes (auto)
CMSIS-DAP Compliant Debugger            yes (auto)
Nu-Link Programmer                      yes (auto)
Cypress KitProg Programmer              yes (auto)
Altera USB-Blaster Compatible           yes (auto)
ASIX Presto Adapter                     yes (auto)
OpenJTAG Adapter                        yes (auto)
SEGGER J-Link Programmer                no
Bus Pirate                              yes (auto)
Use Capstone disassembly framework      no

```

**Reply by Peter-van-Tol** [reply]
I think it is OK, the command is correct.

**Reply by ghost** [reply]
I might just leave it for a day or 2 to see if anything else needs fixing\adding\modding, then do another image.

**Reply by OJthe123** [reply]
The readme tells not to install driver via litexcnc.
what is the waz to update litexcnc when a new release is pushed?

**Reply by ghost** [reply]
pip update should work

Some infos here
https://note.nkmk.me/en/python-pip-usage/

**Comment by ghost** [comment]
Openocd:

For those wanting to install with out having to download a new image this is the package list required:
```
sudo apt-get install libftdi1 libftdi1-2 libftdi1-dev libftdi1-doc libftdi-dev libgpiod-dev libhidapi-dev libhidapi-hidraw0 libhidapi-libusb0 \
libusb-0.1-4 libusb-1.0-0-dev libusb-1.0-doc libusb-dev python3-cffi python3-cffi-backend python3-hidapi python3-ply python3-pycparser
```
Then follow the instructions here:
```
git clone http://openocd.zylin.com/openocd
cd openocd
./bootstrap
./configure --enable-sysfsgpio --enable-bcm2835gpio
make
sudo make install
sudo cp contrib/60-openocd.rules /etc/udev/rules.d/
sudo udevadm --control reload

```
Cleanup:
```sudo apt-get clean```
after everything is verified as working the openocd directory can be deleted. 

**One last thing.**
The openocd binary in oss-cad-suite will need to execute bit removed so that the openocd that was just built & installed gets executed rather than the one in oss-cad-suite.

**Reply by Peter-van-Tol** [reply]
What is the function of
`sudo cp contrib/60-openocd.rules /etc/`

Just wondering, because I've never done that...

**Reply by OJthe123** [reply]
It is like a list of commands that are executed when the system detects a new device that matches i.e. vendor id.
put it in the right groups, set permissions and so on...

**Reply by ghost** [reply]
When a device is plugged in that matches the Vendor ID & Product ID the udev subsystem sets the permissions (group plugdev) can have access to the device without needing to be root.
Once you look into it you can do things like have a certain USB to Serial adapter (matched by serial number if available) come up as the same device name all the time.

More info:
https://opensource.com/article/18/11/udev
https://wiki.archlinux.org/title/udev

Arch wiki always has good info

**Comment by ghost** [comment]
New image
User:cnc
Password:tuxcnc

https://drive.google.com/file/d/1l9dldTeoRwV8Ma3ljQCWFn-8pFGjuIot/view?usp=sharing

**Reply by Peter-van-Tol** [reply]
Would you mind if I put the link to this image in my new project HUB75HAT?

**Reply by OJthe123** [reply]
And can you set or add to the PATH /usr/sbin ?
so one can use dpkg-reconfigure command to change languages 😄 
otherwise it say "command not found", although the package 'debconf' is installed by default.
maybe this has also effects on other pre-installed packages...

**Reply by ghost** [reply]
Something I forgot to mention, the creator of rpi-image-builder includes a utility called menu-config, which allows changing of timezone, language and other bits & bobs.

https://github.com/pyavitz/rpi-img-builder#system-menu-menu-config

**Comment by ghost** [comment]
> Would you mind if I put the link to this image in my new project HUB75HAT?

Not a problem, as it's on just a basic google drive high traffic maybe an issue.
In essence it's for all to use and my way of giving back.

I was thinking a thread dedicated to LiteX-CNC on the Linuxcnc forums might reached a wider audience.

**Reply by Peter-van-Tol** [reply]
Absolutely agree on this. You think open it right away or wait until the HUB75HAT is finished?

**Comment by OJthe123** [comment]
How is openOCD used to flash the card?
I only did it with openFPGALoader.
I have the Altera USB-Blaster

**Reply by Peter-van-Tol** [reply]
I use the following statement with OpenOCD to flash my card:
```shell
openocd \
    -f interface/raspberrypi-native-mod.cfg \
    -c "transport select jtag" \
    -f fpga/lattice_ecp5.cfg \
    -c "init; svf quiet progress colorlight_5a_75e.svf; exit"
```

The statement `-f interface/raspberrypi-native-mod.cfg \` can be replaced with one specific for your Alterra. There is a default configuration available, in which case the command would be
```shell
openocd \
    -f interface/altera-usb-blaster.cfg \
    -c "transport select jtag" \
    -f fpga/lattice_ecp5.cfg \
    -c "init; svf quiet progress colorlight_5a_75e.svf; exit"
```

The structure of the commands is basically:
- select the interface you're going to use, either an Alterrra USB Blaster, Raspberry Pi GPIO or a cheap FT232R dongle;
- select the way to communicate with the FPGA, in this case JTAG;
- define the platform you're going to talk to;
- finally, issue the command to flash the firmware.
